src/ngraph/runtime/gpu/cudnn_emitter.cpp

```diff
@@ -177,6 +178,243 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const runtime::gpu::GPUR
+    auto& filter_desc = get_cudnn_filter_descriptor(filter_shape, data_type, tensor_format);
+    auto& conv_desc = get_cudnn_convolution_descriptor(
+        padding_below, window_movement_strides, window_dilation_strides, mode, data_type);
+
+    const cudnnConvolutionFwdAlgo_t conv_fwd_algo = CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_GEMM;
+    const cudnnConvolutionBwdDataAlgo_t conv_bwd_data_algo = CUDNN_CONVOLUTION_BWD_DATA_ALGO_0;
+    const cudnnConvolutionBwdFilterAlgo_t conv_bwd_filter_algo =
+        CUDNN_CONVOLUTION_BWD_FILTER_ALGO_0;
+
+    size_t workspace_size_in_bytes = 0;
+    switch (direction)
+    {
+    case (Prop::Inference):
+    case (Prop::Forward):
+    {
+        CUDNN_SAFE_CALL(cudnnGetConvolutionForwardWorkspaceSize(*ctx->cudnn_handle,
+                                                                tensor_desc_0,
+                                                                filter_desc,
+                                                                conv_desc,
+                                                                tensor_desc_1,
+                                                                conv_fwd_algo,
+                                                                &workspace_size_in_bytes));
+        break;
+    }
+    case (Prop::Backward):
+    {
+        CUDNN_SAFE_CALL(cudnnGetConvolutionBackwardDataWorkspaceSize(*ctx->cudnn_handle,
+                                                                     filter_desc,
+                                                                     tensor_desc_0,
+                                                                     conv_desc,
+                                                                     tensor_desc_1,
+                                                                     conv_bwd_data_algo,
+                                                                     &workspace_size_in_bytes));
+        break;
+    }
+    case (Prop::BackwardFilter):
+    {
+        CUDNN_SAFE_CALL(cudnnGetConvolutionBackwardFilterWorkspaceSize(*ctx->cudnn_handle,
+                                                                       tensor_desc_0,
+                                                                       tensor_desc_1,
+                                                                       conv_desc,
+                                                                       filter_desc,
+                                                                       conv_bwd_filter_algo,
+                                                                       &workspace_size_in_bytes));
+    }
+    }
+
+    // get an allocator for transient per kernel gpu memory
+    GPUAllocator allocator = this->m_primitive_emitter->get_memory_allocator();
+    // (lazy) allocation for kernel arguments
+    size_t workspace_idx = allocator.reserve_workspace(workspace_size_in_bytes);
+
+    std::unique_ptr<gpu::primitive> conv;
+    switch (direction)
```

Comment:
Convolution is a big enough op that I'm wondering if we should consider breaking its cudnn implementation up into `build_conv_forward` `build_conv_backward` etc. This is fine as is but the switch for the workspace is a bit tedious, and with Prop::BackwardFilter we now have to add that case to future ops.

Reply:
Agree.

Reply:
Let me change it.

Reply:
done